### PR TITLE
Shared memory collective module using One sided MPI communication.

### PR DESCRIPTION
**Pull request by dycz0fx**

### CODE REVIEW

File: ompi/mca/coll/shared/coll_shared_module.c [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+    shared_module->super.coll_module_enable =
+        mca_coll_shared_module_enable;
+    shared_module->super.ft_event = NULL;
+    shared_module->super.coll_allgather = NULL;
+    shared_module->super.coll_allgatherv = NULL;
+    shared_module->super.coll_allreduce = mca_coll_shared_allreduce_intra;      //mca_coll_shared_allreduce_intra
+    shared_module->super.coll_alltoall = NULL;
+    shared_module->super.coll_alltoallv = NULL;
+    shared_module->super.coll_alltoallw = NULL;
+    shared_module->super.coll_barrier = NULL;   //mca_coll_shared_barrier_intra;
```
Comment (thananon) [inline ompi/mca/coll/shared/coll_shared_module.c]: remove comments

File: ompi/mca/coll/shared/coll_shared_module.c [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+        opal_output_verbose(10, ompi_coll_base_framework.framework_output,
+                            "coll:shared:comm_query (%d/%s): intercomm, comm is too small, or not all peers local; disqualifying myself",
+                            comm->c_contextid, comm->c_name);
+        return NULL;
+    }
+
+    /* Get the priority level attached to this module. If priority is less
+     * than or equal to 0, then the module is unavailable. */
+    *priority = mca_coll_shared_component.shared_priority;
+    if (mca_coll_shared_component.shared_priority <= 0) {
```
Comment (thananon) [inline ompi/mca/coll/shared/coll_shared_module.c]: 0 >= mca_coll ..

File: ompi/mca/coll/shared/coll_shared_component.c [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+     /* Initialization / querying functions */
+
+     .collm_init_query = mca_coll_shared_init_query,
+     .collm_comm_query = mca_coll_shared_comm_query,
+     },
+
+    /* shared-component specifc information */
+
+    /* (default) priority */
```
Comment (thananon) [inline ompi/mca/coll/shared/coll_shared_component.c]: remove this

File: ompi/mca/coll/shared/coll_shared_module.c [hunk 4]
```diff
@@ -0,0 +1,346 @@
+/*
```
Comment (thananon) [inline ompi/mca/coll/shared/coll_shared_module.c]: copyright, leave only UT, adjust the year.

File: ompi/mca/coll/shared/coll_shared.h [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+#include "ompi/mca/coll/base/coll_tags.h"
+
+BEGIN_C_DECLS
+    /**
+     * Structure to hold the shared coll component.  First it holds the
+     * base coll component, and then holds a bunch of
+     * shared-coll-component-specific stuff (e.g., current MCA param
+     * values).
+     */
+    typedef struct mca_coll_shared_component_t {
```
Comment (bosilca) [inline ompi/mca/coll/shared/coll_shared.h]: indentation

File: ompi/mca/coll/shared/coll_shared_allreduce.c [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+                                                        sm_ctrl_win);
+    int cur = rank;
+    for (i = 0; i < size; i++) {
+        if (cur != size - 1) {
+            seg_count = l_seg_count;
+        } else {
+            seg_count = count - cur * l_seg_count;
+        }
+        /* Manually call opal_progress to prevent hanging */
+        while (rank != shared_module->ctrl_buf[cur][0]) {
```
Comment (bosilca) [inline ompi/mca/coll/shared/coll_shared_allreduce.c]: Why do we need to call opal_progress here ?

File: ompi/mca/coll/shared/coll_shared_reduce.c [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+    }
+    /* Set up segment count */
+    int seg_count, l_seg_count;
+    seg_count = count / size;
+    l_seg_count = seg_count;
+    if (rank == size - 1) {
+        seg_count = count - rank * l_seg_count;
+    }
+    shared_module->ctrl_buf[rank][0] = rank;
+    shared_module->sm_ctrl_win->w_osc_module->osc_fence(0,
```
Comment (bosilca) [inline ompi/mca/coll/shared/coll_shared_reduce.c]: Without providing an implementation for a barrier the cost of the osc_fence (which is a barrier behind the scene) will be gigantic.

Comment (bosilca) [reply ompi/mca/coll/shared/coll_shared_reduce.c]: We should be able to replace the osc_fence with a counted atomic variable.

File: ompi/mca/coll/shared/coll_shared_module.c [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+    int size = ompi_comm_size(comm);
+    ompi_win_allocate_shared(1 * sizeof(int), sizeof(int),
+                             (opal_info_t *) (&ompi_mpi_info_null), comm,
+                             &shared_module->sm_ctrl_ptr,
+                             &shared_module->sm_ctrl_win);
+    size_t ctrl_size[size];
+    int ctrl_disp[size];
+    shared_module->ctrl_buf = (int **) malloc(sizeof(int *) * size);
+    /* Get ctrl shared memory */
+    for (i = 0; i < size; i++) {
```
Comment (bosilca) [inline ompi/mca/coll/shared/coll_shared_module.c]: Why not allgather ?

**Pull request closed without merging**